File: ml-benchmark/autolifter/single-pass/0after1.f
Config ExtraGrammar = "AutoLifter";
Config SampleIntMin = 0;
Config SampleIntMax = 1;
Config SampleSize = 20;

Inductive List = nil Unit | cons {Int, List};

single_pass = \v: List -> Bool.
  let run = fix (
    \f: List -> Compress List. \xs: List.
    match xs with
      nil _ -> xs
    | cons {h, t} -> cons {h, f t}
    end
  ) in \xs: List.
  v (run xs);

inf = 100;

/*User provided programs*/

zafter1 = fix (
  \f: Bool -> List -> Bool. \seen1: Bool. \xs: List.
  match xs with
    nil _ -> false
  | cons {h, t} ->
    if and seen1 (== h 0) then true
    else f (or seen1 (== h 1)) t
  end
) false;

main = single_pass zafter1;

File: ml-benchmark/autolifter/single-pass/0after1.ml
type list_ = Nil of unit | Cons of int * list_

let single_pass v =
  let rec run xs =
    match xs with
    | Nil _ -> xs
    | Cons (h, t) ->
      let t2 = run t in
      Cons (h, t2)
  in
  fun xs -> v (run xs)

let inf = 100

let zafter1 =
  let rec f seen1 xs =
    match xs with
    | Nil _ -> false
    | Cons (h, t) ->
      if seen1 && h = 0 then true
      else f (seen1 || h = 1) t
  in
  fun xs -> f false xs

let program = single_pass zafter1